packages/kbn-config-schema/src/index.ts

```diff
@@ -100,6 +101,10 @@ function maybe<V>(type: Type<V>): Type<V | undefined> {
   return new MaybeType(type);
 }
 
+function nullable<V>(type: Type<V>): Type<V | null> {
+  return new NullableType(type);
```

Comment:
__question:__ I'm curious what didn't you want to implement this like you've already done in your alerting code? Are there any issues with that?
```ts
function nullable<V>(type: Type<V>): Type<V | null> {
  return oneOf([type, literal(null)], { defaultValue: null });
}
```

Reply:
The way I implemented it in user-land code (the alerting code) was the only way I could really do it, as far as I know.

The implementation here was copied from the `maybe` implementation, changing `undefined` to `null`, which seemed like a more straightforward fashion, and fit in with the existing framework, working like an existing composite type.

One of the downsides of the implementation I did in the alerting code is that an error message that doesn't meet the `V` constraints, was a 3 line error message, something like:

```
xyz did not meet constraints:
- it wasn't a V
- it wasn't null
```

Guessing that came from the `oneOf` implementation.

Reply:
when I run a snippet 
```js
const type = schema.nullable(schema.string());
expect(type.validate(1)).toEqual(null);
```
with implementation from Oleg I see:
```
types that failed validation:
  - [0]: expected value of type [string] but got [number]
  - [1]: expected value to equal [null] but got [1]
```

with implementation from this PR:
```
 expected value of type [string] but got [number]
```
which is not fully correct, because `null` is a valid value as well

Reply:
also maybe is a bit different. usually `undefined` is not valid value, unlike `null`.

Reply:
I get the same result for:

```
const type = schema.maybe(schema.string());
expect(type.validate(1)).toEqual(null);
```

I'm not sure that extra precision is worth it.  Certainly not if the message extends to > 1 line.

Reply:
Do we want to explore another way to solve this?  It seems like easily allowing `null` values - as an opt-in - is something other folks may want, beyond just `undefined` (via `maybe()`).  In previous cases where I've used `joi`  (not at Elastic), I've done this via `allow(null)`, which ... is what I'm using here as well:

https://github.com/elastic/kibana/blob/a020c3c674f74339c69054a51b6171295104b155/packages/kbn-config-schema/src/types/nullable_type.ts#L25-L29

I only know about this because I **HAVE** needed it in the past :-)

The current implementation in this PR seems like the most straightforward way to do it, to me, right now - but I would not be surprised to find if there's a better way!

How about extending `maybe()` with an option to allow `null` as well as `undefined`?  I noticed `maybe()` doesn't take options like many of the other schema types.  But it also seems like it would be not quite right to extend the `maybe` type at this point, from:

```
function maybe<V>(type: Type<V>): Type<V | undefined>
```

to (adding some options and the `null` result)

```
function maybe<V>(type: Type<V>, options?: MaybeOptions): Type<V | undefined | null>
```

Could break some existing clients, to have that new `null` alternative result?

Reply:
for me
`undefined` - unintentional absence of value
`null` -  intentional absence of value

imo it's a valid behaviour when `Maybe` doesn't offer to set `undefined` as a config value.

> How about extending maybe() with an option to allow null as well as undefined?

for me, it creates more questions - should the library falls back to defaultValue if `schema.validate(null)`? how plugin code should treat a difference between `null` and `undefined`? what if their logic doesn't support `null` as a valid value?

Reply:
I think the only way to get the information re: `null` included in the error message is to use `oneOf()`, and I'm not really interested in that due to the verbosity of the error messages.

For example, in the example above of: `schema.nullable(schema.string()).validate(1)`, the message seems to be generated here, so there's not a great option of reusing those messages and inserting a `null` in the message.

https://github.com/elastic/kibana/blob/88fffd5f642cf7c34af6e1af4333e0de2e111c7d/packages/kbn-config-schema/src/types/string_type.ts#L53